src/history/HistoryTests.cpp

```diff
@@ -204,6 +225,145 @@ TEST_CASE("History publish queueing", "[history][historydelay][historycatchup]")
           catchupSimulation.getApp().getLedgerManager().getLedgerNum());
 }
 
+TEST_CASE("History bucket verification",
+          "[history][bucketverification][batching]")
+{
+    /* Tests bucket verification stage of catchup. Assumes ledger chain
+     * verification was successful. **/
+
+    Config cfg(getTestConfig());
+    VirtualClock clock;
+    auto cg = std::make_shared<TmpDirHistoryConfigurator>();
+    cg->configure(cfg, true);
+    Application::pointer app = createTestApplication(clock, cfg);
+    CHECK(app->getHistoryArchiveManager().initializeHistoryArchive("test"));
+
+    auto bucketGenerator = TestBucketGenerator{
+        *app, app->getHistoryArchiveManager().getHistoryArchive("test")};
+    std::vector<std::string> hashes;
+    auto& wm = app->getWorkManager();
+    std::map<std::string, std::shared_ptr<Bucket>> mBuckets;
+    auto tmpDir =
+        std::make_unique<TmpDir>(app->getTmpDirManager().tmpDir("bucket-test"));
+    hashes.push_back(
+        bucketGenerator.generateBucket(TestBucketState::CONTENTS_AND_HASH_OK));
+
+    SECTION("successful download and verify")
+    {
+        hashes.push_back(bucketGenerator.generateBucket(
+            TestBucketState::CONTENTS_AND_HASH_OK));
+        auto verify =
+            wm.executeWork<DownloadBucketsWork>(mBuckets, hashes, *tmpDir);
+        REQUIRE(verify->getState() == Work::WORK_SUCCESS);
+    }
+    SECTION("download fails file not found")
+    {
+        hashes.push_back(
+            bucketGenerator.generateBucket(TestBucketState::FILE_NOT_UPLOADED));
+
+        auto verify =
+            wm.executeWork<DownloadBucketsWork>(mBuckets, hashes, *tmpDir);
+        REQUIRE(verify->getState() == Work::WORK_FAILURE_RAISE);
+        downloadStatusCheck(*verify, false);
+    }
+    SECTION("download succeeds but unzip fails")
+    {
+        hashes.push_back(bucketGenerator.generateBucket(
+            TestBucketState::CORRUPTED_ZIPPED_FILE));
+
+        auto verify =
+            wm.executeWork<DownloadBucketsWork>(mBuckets, hashes, *tmpDir);
+        REQUIRE(verify->getState() == Work::WORK_FAILURE_RAISE);
+        downloadStatusCheck(*verify, false);
+    }
+    SECTION("verify fails with duplicate buckets")
+    {
+        hashes.push_back(bucketGenerator.generateBucket(
```

Comment:
per our chat, duplicate buckets and out of order buckets don't need to be tested here (we can rely 100% on hash and the fact that the overall network is well behaved from that point of view)